### PR TITLE
Update viewport scaling demo to 4.0.

### LINKED ISSUE #856: 3D Scaling in 4.0-dev not working
**Which demo project is affected:**
3D scaling in viewport

**OS/device including version:**
Linux Arch lts

**Issue description:**
Filtering works, however the scaling does not work at all. I've been trying to fix this demo myself as I need it for some of my personal projects but I can't find what's wrong with it. Is this still a 4.0 bug or is there something I'm missing here? Setting the size of the subviewport does not work and the size 2D override doesn't seem to be doing anything.

### PR DESCRIPTION
**Pull request by rcorre**
Fixes #856.

The 3d_scaling demo was updated to use the built-in 3D scaling mechanisms
in the Forward+ renderer.

A new 3d_scaling_compat demo was added to demonstrate the legacy
SubViewport-based scaling method for the Compatibility renderer.

### CODE REVIEW

File: viewport/3d_scaling_compat/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+be used to have a "pixel art" viewport for the main game area and a
+non-pixel-art viewport for HUD elements.
+
+If you're using the `Forward+` renderer, see ../3d_scaling instead.
+
+Language: GDScript
+
+Renderer: Compatibility
+
+Check out this demo on the asset library: https://godotengine.org/asset-library/asset/586
```
Comment (rcorre) [inline viewport/3d_scaling_compat/README.md:15]: Do we need a new asset library entry?

Comment (Calinou) [reply viewport/3d_scaling_compat/README.md:15]: Yes. 4.0 demos haven't been submitted for the asset library yet, but we intend to keep the 3.5 demos available for the foreseeable future.

**Comment by rcorre** [comment]
> but the 3D Resolution Scaling (Compatibility) demo needs to resize the SubViewport when the window size changes

Huh, I figured that would happen automatically with **Stretch** enabled. I added this print:

```
get_tree().root.size_changed.connect(func(): prints(get_tree().root.size, viewport_container.size, viewport.size))
```

The viewport is resizing to the parent control, but the way the parent control resizes depends on the `window/stretch/mode` and `window/stretch/aspect` settings in the project. If you set `mode=disabled`, then the viewport container and the viewport auto-adjust to the window size.

**Comment by leetNightshade** [comment]
Idk if it's just me, but testing this on Godot 4.1.1.stable.custom_build on Steam Deck doesn't seem to do anything, it draws at the window resolution. I have this same problem in my project that emulates this subviewport project.

File: viewport/3d_scaling/hud.gd [hunk 2]
```diff
@@ -1,42 +1,25 @@
 extends Control
 
-# The 3D viewport's scale factor. For instance, 1.0 is full resolution,
-# 0.5 is half resolution and 2.0 is double resolution. Higher values look
-# sharper but are slower to render. Values above 1 can be used for supersampling
-# (SSAA), but filtering must be enabled for supersampling to work.
-var scale_factor = 1.0
+# The 3D viewport's shrink factor. For instance, 1 is full resolution,
```
Comment (AThousandShips) [inline viewport/3d_scaling/hud.gd:3]:
```suggestion

# The 3D viewport's shrink factor. For instance, 1 is full resolution,
```

File: viewport/3d_scaling/hud.gd [hunk 3]
```diff
@@ -1,42 +1,25 @@
...
+var scale_factor = 1
+var filter_mode = Viewport.SCALING_3D_MODE_BILINEAR
 
-@onready var viewport_container = $SubViewportContainer
-@onready var viewport = $SubViewportContainer/SubViewport
+@onready var viewport = get_tree().root
 @onready var scale_label = $VBoxContainer/Scale
 @onready var filter_label = $VBoxContainer/Filter
 
 func _ready():
```
Comment (AThousandShips) [inline viewport/3d_scaling/hud.gd:13]:
```suggestion

func _ready():
```

File: viewport/3d_scaling_compat/hud.gd [hunk 4]
```diff
@@ -0,0 +1,26 @@
+extends Control
+
+# The 3D viewport's shrink factor. For instance, 1 is full resolution,
```
Comment (AThousandShips) [inline viewport/3d_scaling_compat/hud.gd:3]:
```suggestion

# The 3D viewport's shrink factor. For instance, 1 is full resolution,
```

**Comment by Calinou** [comment]
https://github.com/godotengine/godot/pull/83976 is landing in 4.3, so I'd prefer we just use the 3D scaling method for all rendering methods now.

**Comment by rcorre** [comment]
I updated to only include the builtin scaling demo, and removed the 3D compat demo.
I'm not as familiar with 2D, so I didn't attempt to create a 2D scaling demo.

**Comment by Calinou** [comment]
Merged manually with https://github.com/godotengine/godot-demo-projects/commit/722bd1168955b25e498d9b8967caf3970e02877e (which includes some visual and UI improvements). Thanks!

Screenshot

**Pull request closed without merging**